core/src/main/java/org/infinispan/xsite/statetransfer/SyncProviderState.java

```diff
@@ -0,0 +1,239 @@
+   }
+
+   @Override
+   public void cancelTransfer() {
+      OutboundTask currentTask = TASK_UPDATER.getAndSet(this, null);
+      if (currentTask != null) {
+         currentTask.cancel();
+      }
+   }
+
+   @Override
+   public boolean isSending() {
+      return task != null;
+   }
+
+   @Override
+   public boolean isOriginatorMissing(Collection<Address> members) {
+      OutboundTask currentTask = task;
+      return currentTask != null && !members.contains(currentTask.coordinator);
+   }
+
+   // methods for OutboundTask
+   void taskFinished() {
+      TASK_UPDATER.set(this, null);
+   }
+
+   XSiteBackup getBackup() {
+      return backup;
+   }
+
+   int getChunkSize() {
+      return configuration.chunkSize();
+   }
+
+   long getWaitTimeMillis() {
+      return configuration.waitTime();
+   }
+
+   int getMaxRetries() {
+      return configuration.maxRetries();
+   }
+
+   private static class OutboundTask implements XSiteStatePushTask, Predicate<List<XSiteState>>, Function<List<XSiteState>, CompletableSource>, CompletableObserver {
+
+      private final Address coordinator;
+      private final XSiteStateProvider provider;
+      private final SyncProviderState state;
+      private volatile boolean canceled = false;
+
+      private OutboundTask(Address coordinator, XSiteStateProvider provider, SyncProviderState state) {
+         this.coordinator = coordinator;
+         this.provider = provider;
+         this.state = state;
+      }
+
+      @Override
+      public void execute(Flowable<XSiteState> flowable, CompletionStage<Void> delayer) {
+         //delayer is the cache topology future. we need to ensure the topology id is installed before iterating
+         delayer.thenRunAsync(() -> flowable
+                     .buffer(state.getChunkSize())
+                     .takeUntil(this)
+                     .concatMapCompletable(this, 1)
+                     .subscribe(this),
+               provider.getExecutor());
+
+      }
+
+      public void cancel() {
+         canceled = true;
+      }
+
+      @Override
+      public boolean test(List<XSiteState> ignored) {
+         //Flowable#takeUntil method
+         return canceled;
+      }
+
+      @Override
+      public void onSubscribe(@NonNull Disposable d) {
+
+      }
+
+      @Override
+      public void onComplete() {
+         //if canceled, the coordinator already cleanup the resources. There is nothing to be done here.
+         if (canceled) {
+            return;
+         }
+         provider.notifyStateTransferEnd(state.getBackup().getSiteName(), coordinator, true);
+         state.taskFinished();
+      }
+
+      @Override
+      public void onError(@NonNull Throwable e) {
+         //if canceled, the coordinator already cleanup the resources. There is nothing to be done here.
+         if (canceled) {
+            return;
+         }
+         provider.notifyStateTransferEnd(state.getBackup().getSiteName(), coordinator, false);
```

Comment:
Is this throwable logged somewhere?

Reply:
no, I'll add it to the `CommandRetry` class.